migrate code base to media-query-list-parser

romainmenke opened the pull request:
see : 
- https://github.com/postcss/postcss-media-minmax/issues/42
- https://github.com/csstools/postcss-plugins/tree/postcss-preset-env--v8/packages/media-query-list-parser#readme

summary :
- travis -> github workflow
- node 10 -> node 14
- javascript -> typescript
- regexp -> media-query-list-parser

Our media query list parser has a very large OM and it is a low level API.
Working with this in plain JavaScript is really painful. It is much easier to work with from TypeScript.

To support TypeScript I've added the build config (rollup) from the csstools mono repo.
There are other ways to get a similar result, but this is one I am comfortable with.

This pull request fixes a few issues but mostly focusses on using the new parser and OM to achieve the same outcome.

-----

note : 

Totally fine if this is not a direction you want to take with this plugin.
What works for us doesn't necessarily work for you :)

romainmenke committed:
update lib

romainmenke committed:
add more tests

romainmenke committed:
fix typo

romainmenke commented:
@yisibl Do you have time in the next few weeks to look at this?

We are hoping to wrap up all work for `postcss-preset-env` version 8 and would like to include an updated version of this plugin.

As I said before we totally understand it if you want to go in another direction.

yisibl commented:
Thank you for your contribution, I need to chat with my partner and get back to you.

romainmenke commented:
@yisibl Any news here?

yisibl commented:
Thank you for your continued interest, I should have time during the Chinese New Year to carry out my work here.

yisibl commented on test/fixtures/color-index.output.css:
```diff
@@ -1,5 +1,5 @@
 @media screen and (min-color-index: 0) and (max-color-index: 8) {}
 
-@media screen and (min-color-index: 0) and (max-color-index: 8) {}
+@media screen and (min-color-index:0 ) and (max-color-index: 8) {}
```
Is it possible to not output extra spaces here `(min-color-index:0 )`?

romainmenke replied:
This is the space that was originally placed after `0` in something like `(0 < color-index < 8)`

I've been considering adding a serializer function to the package based on : https://www.w3.org/TR/cssom-1/#serializing-media-queries

Would this be a good solution?

Or is it preferable to have some heuristics that try to preserve the original code style/format?

romainmenke replied:
> I've been considering adding a serializer function to the package based on

This is not interesting because CSSOM serializing algorithms do not produce equivalent CSS source code. Things like comments will be omitted.

I've update the code to flip `before` and `after` whitespace.

romainmenke replied:
Also changed a few other bits.
There no longer are any differences in existing tests.

romainmenke committed:
update whitespace and comment handling

romainmenke committed:
more output formatting

romainmenke committed:
more whitespace handling

romainmenke commented on index.js:
```diff
@@ -1,122 +0,0 @@
-    media: (atRule) => {
-      transform(atRule);
-    },
-    'custom-media': (atRule) => {
-      transform(atRule);
-    },
```
`custom-media` is no longer handled.
I can bring it back, but not sure if this is an interesting feature.

No browser supports `@custom-media`, so needs to be transpiled to regular `@media` anyway.

Upcoming release of our media query list parser will include support for `@custom-media` parsing, so if it should be implemented it will not be a major effort.

romainmenke replied:
`custom-media` parsing was added to the media query list parser.

This feature has been restored.

romainmenke commented:
Related issues that will likely be fixed by this change : 
- https://github.com/postcss/postcss-media-minmax/issues/38
- https://github.com/postcss/postcss-media-minmax/issues/34
- https://github.com/postcss/postcss-media-minmax/issues/19
- support for `<ratio>` in queries that need altered values (`@media (0/0 < aspect-ratio < 16/9)`)

romainmenke committed:
illustrate ratio issue

romainmenke committed:
cleanup

romainmenke commented:
@yisibl Any chance you could give this another look?
Again totally fine if this parser is not something you want to adopt but we really want to ship an improved version for `postcss-preset-env`. We also don't mind maintaining a fork but we prefer to contribute here.

_The parser and underlying tokenizer have been in use now for months in both `postcss-preset-env` and `stylelint` without any issues._

yisibl commented:
@romainmenke Sorry, too busy with work this month, I'm afraid I'll have time next month!

romainmenke committed:
bring over latest patches

romainmenke commented:
That is fine @yisibl :)

We really want to move forward with this, given the momentum this feature currently has in browsers.

I've created a new plugin based on this PR in the csstools mono repo.
I've also added a bunch more tests, including browser tests.

Ideally these patches find there way back to this plugin in some form or another after which we can delete the forked plugin and set `postcss-preset-env` right again :)

I've already updated this PR with the latest bug fixes.
I haven't updated the PR yet that doesn't include TypeScript.
